### ISSUE TITLE
High CPU usage caused by "docker stats"

### ISSUE THREAD

**Issue by Programie** [moby/moby#39664]
**Description**

I've a standalone Docker host (i.e. no swarm) on which 14 containers are running.

Whenever I start `docker stats` the CPU usage goes up from almost zero CPU usage to about 15 to 20% (caused by the processes containerd-shim, containerd and dockerd). Once I stop `docker stats`, the CPU usage goes back to normal (almost zero).

I wanted to use the [Docker collectd plugin](https://github.com/signalfx/docker-collectd-plugin) which uses the stats call of the Docker API in the background and thus having the same impact on the CPU usage. But with that huge increase in CPU usage, I better do not collect Docker container metrics.

I also tested to get some metrics directly from `/sys/fs/cgroup/memory/docker` (for memory stats). Executing something like `watch -n 1 cat /sys/fs/cgroup/memory/docker/*/memory.stat` doesn't have any visible impact on the CPU usage, so I guess it's Docker doing something in the background which causes the high CPU usage.

**Steps to reproduce the issue:**
1. Start some containers
2. Check the current CPU usage: should be almost zero (if nothing else is using the resources)
3. Execute `docker stats` and keep it running
3. Check the current CPU usage: should be increased depending on the number of containers currently running

**Describe the results you received:**

The CPU usage increases by about 15 to 20%.

**Describe the results you expected:**

The CPU usage should not increase that much.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.1
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        74b1e89
 Built:             Thu Jul 25 21:22:03 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.1
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       74b1e89
  Built:            Thu Jul 25 21:20:35 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 16
  Running: 14
  Paused: 0
  Stopped: 2
 Images: 16
 Server Version: 19.03.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.9.0-9-amd64
 Operating System: Debian GNU/Linux 9 (stretch)
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 7.615GiB
 Name: *REMOVED*
 ID: KBZ2:L5HD:PVYJ:OY5V:LH3O:4SCD:QQ65:NYVZ:Z6VX:WC7B:ZY3C:6VU4
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Docker is running on Hetzner Cloud server (CX31).

**Comment by danez** [comment]
I know this is old, but this started happening for me too when i switched to containerd-image-storage. On my Raspberri Pi 5, during docker stats, cpu goes up to nearly 100%.

**Comment by Programie** [comment]
I checked it again with the Docker version I'm currently using (28.5.1) but could not reproduce it anymore. In my case it might got fixed in the last few years since reporting the issue even if it wasn't closed by the developers.

I also checked it on another server which is already running Docker 29 (29.1.2) and also could not reproduce the issue anymore.

I also saw that Docker 29 switched to containerd image store by default. So, that should not be the problem. Maybe it's not optimized for ARM as I'm using x86 hardware without having that issue.

**Comment by thaJeztah** [comment]
Yeah, ISTR there was a fix at some point to prevent a tight loop when collecting stats, and I _suspect_ that was the original issue.


@danez if you have more details, perhaps it's better to open a new ticket with details output your environment (at least full `docker version` and `docker info` output; if there's any (system) logs for containerd and/or docker that may be of interest.
